File: src/com/home/Chapter_5_Linked_Lists/SortedListApp/SortedList.java
package com.home.Chapter_5_Linked_Lists.SortedListApp;

/**
 * Created by Home on 08.12.2016.
 */
public class SortedList {
    private Link first;

    public SortedList(){
        first = null;
    }

    public boolean isEmpty(){
        return (first == null);
    }

    public void insert(long key){
        Link newLink = new Link(key);
        Link previous = null;
        Link current = first;

        while(current != null && key > current.dData){
            previous = current;
            current = current.next;
        }
        if(previous == null){
            first = newLink;
        } else {
            previous.next = newLink;
        }
        newLink.next = current;
    }

    public Link remove(){
        Link temp = first;
        first = first.next;
        return temp;
    }

    public void displayList(){
        System.out.println("List (first-->last): ");
        Link current = first;
        while(current != null){
            current.displayLink();
            current = current.next;
        }
        System.out.println("");
    }
}

File: src/com/home/Chapter_11_Hash_Tables/hashChain/SortedList.kt
package com.home.Chapter_11_Hash_Tables.hashChain

class SortedList(var first: Link? = null) {

    fun insert(link: Link) {
        val key = link.iData
        var previous: Link? = null;
        var current: Link? = first

        while (current != null && key > current.iData) {
            previous = current
            current = current.next
        }
        if (previous == null) {
            first = link
        } else {
            previous.next = link
        }
        link.next = current
    }

    fun delete(key: Int) {
        var previous: Link? = null;
        var current: Link? = first

        while (current != null && key != current.iData) {
            previous = current
            current = current.next
        }

        if (previous == null) {
            first = first?.next
        } else {
            previous.next = current?.next
        }
    }

    fun find(key: Int): Link? {
        var current: Link? = first
        while (current != null && current.iData <= key) {
            if (current.iData == key) return current
            current = current.next
        }
        return null
    }

    fun displayList() {
        print("List (first --> last): ")
        var current: Link? = first
        while (current != null) {
            current.display()
            current = current.next
        }
        println("")
    }

}